File: frontend/src/causal.ts
// 인과 경로 응답 → Cytoscape elements. **순수 함수다** — Cytoscape를 import하지 않는다.
//
// **한 주가 한 판이다.** 상세가 그 주의 경로 전부를 주므로, 사건들과 대상들이 채널을 거쳐
// 이어지는 모양을 그대로 그린다. 경로 하나만 그리면 직선 하나라 그림이 말해 주는 것이 없다.
//
// **다중 홉은 노드를 공유해서 생긴다.** 대상에서 출발한 경로의 원인 노드는 다른 경로의
// 대상 노드와 **같은 id**다 — `VIX → NASDAQ100_FUT → SOX → 005930`이 그렇게 이어진다.
// 출발점을 새 사건으로 만들었다면 `target:SOX`와 `event:SOX 상승`이 갈려 사슬이 끊긴다.
//
// **채널 노드는 이름으로 공유한다.** 두 경로가 `할인율`을 함께 거쳤다는 사실이 요점이고,
// 경로마다 노드를 새로 만들면 그 공유가 사라진다.

import type { CytoscapeElement } from "./graph";
import { CAUSAL_TARGET_KINDS, labelOf } from "./labels";
import type { CausalGraph, CausalPathRow } from "./types";

/** 사건 노드 id. 한 주에 사건이 여럿이라 id에 사건 번호가 들어간다. */
export function eventId(row: CausalPathRow): string {
  return `event:${row.event_id}`;
}

/** 채널 노드 id. 이름이 곧 정체성이라 이름으로 만든다. */
export function channelId(name: string): string {
  return `channel:${name}`;
}

/** 대상 노드 id. 같은 코드라도 종류가 다르면 다른 대상이다(`US10Y`는 시세, `KTB10Y`는 지표). */
export function targetId(kind: string, code: string): string {
  return `target:${kind}:${code}`;
}

/**
 * 이 경로의 출발 노드.
 *
 * 대상 출발이면 **대상 노드와 같은 id**를 낸다 — 그 공유가 다중 홉의 전부다.
 */
export function sourceId(row: CausalPathRow): string {
  return row.source_kind === "target" && row.source_target_code !== null
    ? targetId(row.source_target_kind ?? "quote", row.source_target_code)
    : eventId(row);
}

/** 경로 하나가 지나는 노드 순서. 출발에서 대상까지다. */
export function walk(row: CausalPathRow): string[] {
  return [
    sourceId(row),
    ...row.channels.map(channelId),
    targetId(row.target_kind, row.target_code),
  ];
}

/** 이 경로의 엣지 id들. 강조와 중복 제거가 같은 키를 봐야 한다. */
export function edgeIds(row: CausalPathRow): string[] {
  const nodes = walk(row);
  return nodes.slice(0, -1).map((from, index) => `${from}->${nodes[index + 1]}`);
}

/** 대상 노드의 라벨. 코드 아래에 종류를 적어 `US10Y`가 시세인지 지표인지 밝힌다. */
function targetLabel(kind: string, code: string): string {
  return `${code}\n${labelOf(CAUSAL_TARGET_KINDS, kind)}`;
}

/**
 * 한 주의 그래프.
 *
 * `picked`는 지금 보고 있는 경로다 — 그 경로의 엣지만 방향 색을 입힌다. 전부에 색을 주면
 * 한 화면에서 위·아래가 뒤섞여 어느 것이 이 경로의 주장인지 사라진다.
 */
export function causalElements(
  paths: readonly CausalPathRow[],
  picked: CausalPathRow | null,
): CytoscapeElement[] {
  if (paths.length === 0) return [];

  const nodes = new Map<string, CytoscapeElement>();
  const edges = new Map<string, CytoscapeElement>();
  const highlighted = new Set(picked === null ? [] : edgeIds(picked));
  // 지금 경로의 출발 노드만 가운데 표시를 받는다. 한 주에 사건이 여럿이라 전부 키우면
  // 강조가 강조가 아니게 된다.
  const center = picked === null ? null : sourceId(picked);

  const putTarget = (kind: string, code: string) => {
    const id = targetId(kind, code);
    // 이미 있으면 덮어쓰지 않는다 — 같은 노드를 다른 경로가 결과로도 원인으로도 쓴다.
    if (!nodes.has(id)) {
      nodes.set(id, {
        group: "nodes",
        data: { id, label: targetLabel(kind, code) },
        classes: id === center ? "target center" : "target",
      });
    }
  };

  for (const row of paths) {
    if (row.source_kind === "event") {
      const id = eventId(row);
      nodes.set(id, {
        group: "nodes",
        data: { id, label: row.event_title ?? id },
        classes: id === center ? "event center" : "event",
      });
    } else if (row.source_target_code !== null) {
      putTarget(row.source_target_kind ?? "quote", row.source_target_code);
    }

    for (const name of row.channels) {
      nodes.set(channelId(name), {
        group: "nodes",
        data: { id: channelId(name), label: name },
        classes: "channel",
      });
    }
    putTarget(row.target_kind, row.target_code);

    const chain = walk(row);
    chain.slice(0, -1).forEach((from, index) => {
      const to = chain[index + 1]!;
      const key = `${from}->${to}`;
      // **엣지는 한 번만 만든다.** 두 경로가 같은 두 노드를 잇는 일이 흔하고, 그때
      // 겹쳐 그리면 화살표가 굵어질 뿐 새 사실이 없다.
      if (edges.has(key) && !highlighted.has(key)) return;
      edges.set(key, {
        group: "edges",
        data: { id: key, source: from, target: to, label: "" },
        classes: highlighted.has(key) ? `cites direction-${picked?.sign ?? "up"}` : "cites",
      });
    });
  }

  return [...nodes.values(), ...edges.values()];
}

/**
 * **그래프 DB가 준 투영**을 그대로 그린다.
 *
 * 경로 목록에서 조립하는 `causalElements`와 다른 점은 하나다 — 투영은 노드를 **주를 넘어**
 * 공유하므로 08-10 주에 닿은 `SOX`가 08-17 주의 원인으로 이어지는 사슬이 함께 온다.
 * 그 사슬이 Neo4j를 들인 이유이고(4단계 설계 §1), 조회가 시각 역행을 이미 걸러 준다.
 *
 * `pickedPathId`가 있으면 그 경로의 엣지만 방향 색을 받는다. 색을 전부에 주면 한 화면에서
 * 위·아래가 뒤섞여 어느 것이 이 경로의 주장인지 사라진다.
 */
export function graphElements(
  graph: CausalGraph,
  pickedPathId: number | null,
  pickedSign: string = "up",
): CytoscapeElement[] {
  const nodes = graph.nodes.map((node) => ({
    group: "nodes" as const,
    data: { id: node.id, label: node.label },
    classes: node.kind,
  }));

  const edges = new Map<string, CytoscapeElement>();
  for (const edge of graph.edges) {
    const mine = pickedPathId !== null && edge.path_id === pickedPathId;
    const key = `${edge.source}->${edge.target}`;
    // 엣지는 한 번만 그린다. 여러 경로가 같은 두 노드를 잇는 일이 흔하고, 그때 겹쳐
    // 그리면 화살표가 굵어질 뿐 새 사실이 없다. 다만 강조는 덮어쓴다.
    if (edges.has(key) && !mine) continue;
    edges.set(key, {
      group: "edges",
      data: { id: key, source: edge.source, target: edge.target, label: "" },
      classes: mine ? `cites direction-${pickedSign}` : "cites",
    });
  }

  return [...nodes, ...edges.values()];
}

File: apps/api/service/causal.py
"""인과 그래프의 매핑.

**경로 한 줄이 사건·체인·대상·실현 등락을 다 들고 나간다.** 화면이 단계마다 다시 묻지
않게 하는 것이 이 층의 일이고, 리포지토리가 준 행 묶음을 자리로 푸는 것도 여기서 한다.
"""

from datetime import date
from typing import Any

from apps.api.repository import (
    DEFAULT_LIMIT,
    CausalGraphReadRepository,
    GraphRows,
    MarketCausalReadRepository,
    PathRows,
)
from apps.api.schemas import (
    CausalChannelList,
    CausalChannelRow,
    CausalEventList,
    CausalEventRow,
    CausalEvidenceRow,
    CausalGraph,
    CausalGraphEdge,
    CausalGraphNode,
    CausalPathDetail,
    CausalPathList,
    CausalPathRow,
)
from apps.api.service.common import dart_url, number


def _text(value: Any) -> str:
    """Enum이면 값을, 아니면 문자열을."""
    return str(getattr(value, "value", value))


def path_of(path: Any, event: Any | None, channels: tuple[str, ...]) -> CausalPathRow:
    """**사건이 없는 경로가 정상이다.** 대상에서 출발한 경로는 `event_id`가 NULL이고
    `source_target_*` 셋이 채워진다 — 그것을 버리면 다중 홉의 절반이 화면에서 사라진다."""
    return CausalPathRow(
        id=path.id,
        week_start=path.week_start,
        source_kind="event" if path.event_id is not None else "target",
        event_id=path.event_id,
        event_title=None if event is None else event.title,
        event_occurred_on=None if event is None else event.occurred_on,
        source_target_kind=None
        if path.source_target_kind is None
        else _text(path.source_target_kind),
        source_target_code=path.source_target_code,
        source_sign=None if path.source_sign is None else _text(path.source_sign),
        target_kind=_text(path.target_kind),
        target_code=path.target_code,
        channels=channels,
        sign=_text(path.sign),
        confidence=_text(path.confidence),
        reasoning=path.reasoning,
        return_week_change=number(path.return_week_change) or 0.0,
        return_t1_change=number(path.return_t1_change) or 0.0,
        return_t5_change=number(path.return_t5_change) or 0.0,
        return_unit=_text(path.return_unit),
        llm_run_id=path.llm_run_id,
    )


def build_paths(rows: PathRows, *, limit: int, offset: int) -> CausalPathList:
    """**모든 경로를 낸다.** 전에는 사건을 못 찾은 행을 버렸는데, 출발점이 대상인 경로가
    생기면서 그 규칙이 그것들을 통째로 삼켰다(2026-08-30). 사건 출발인데 사건이 없는 것은
    외래키가 막고 있다."""
    return CausalPathList(
        items=tuple(
            path_of(path, rows.events.get(path.id), rows.chains.get(path.id, ()))
            for path in rows.paths
        ),
        limit=limit,
        offset=offset,
        has_more=rows.has_more,
    )


class UnknownPath(Exception):
    """없는 경로 id. 라우트가 404로 바꾼다."""


class GraphOffline(Exception):
    """그래프 DB가 이 실행에 붙어 있지 않다. 라우트가 503으로 바꾼다.

    **빈 그래프로 위장하지 않는다** — "그 주에 경로가 없다"와 "그래프 DB가 꺼져 있다"는
    화면이 다르게 말해야 한다. 화면은 이 응답을 보고 경로 목록으로 그림을 조립한다.
    """


def graph_of(rows: GraphRows, week: date | None) -> CausalGraph:
    """투영을 응답 계약으로. **모양만 옮긴다** — 숫자는 경로 응답이 갖는다."""
    return CausalGraph(
        source="neo4j",
        week_start=week,
        nodes=tuple(
            CausalGraphNode(id=node.id, kind=node.kind, label=node.label) for node in rows.nodes
        ),
        edges=tuple(
            CausalGraphEdge(
                source=edge.source,
                target=edge.target,
                type=edge.type,
                path_id=edge.path_id,
                week_start=edge.week_start,
                position=edge.position,
            )
            for edge in rows.edges
        ),
    )


def evidence_of(path_id: int, ref: str, titles: dict[int, str]) -> CausalEvidenceRow:
    """`<kind>:<id>` 규약을 응답에서 한 번만 푼다.

    **화면이 문자열을 쪼개게 두지 않는다** — 그러면 규약이 두 곳에 생기고, 종류가 늘 때
    한쪽만 고친다. 모르는 종류는 종류 이름만 남기고 링크를 만들지 않는다.
    """
    kind, _, value = ref.partition(":")
    title = titles.get(int(value)) if kind == "document" and value.isdigit() else None
    url = None
    if kind == "document" and value.isdigit():
        # 문서는 우리 화면에 상세가 있다. 원문 링크는 그 상세가 준다.
        url = f"/documents/{value}"
    elif kind == "disclosure" and value:
        # 공시는 접수번호가 곧 주소다. 이 근거는 DART에서 온 것만 있다.
        url = dart_url("dart", value)
    return CausalEvidenceRow(path_id=path_id, ref=ref, kind=kind, title=title, url=url)


def build_detail(rows: PathRows, path_id: int) -> CausalPathDetail:
    """**그 주 전체를 함께 낸다.** 경로 하나만 내면 화면이 그릴 것이 직선 하나뿐이고,
    사건 단위로 묶으면 대상이 다시 원인이 되는 사슬이 끊어진 채로 보인다."""
    family = tuple(
        path_of(path, rows.events.get(path.id), rows.chains.get(path.id, ()))
        for path in rows.paths
    )
    found = next((row for row in family if row.id == path_id), None)
    if found is None:
        raise UnknownPath(str(path_id))
    return CausalPathDetail(
        path=found,
        siblings=family,
        evidence=tuple(
            evidence_of(path, ref, rows.document_titles) for path, ref in rows.evidence
        ),
    )


def event_of(row: tuple[Any, ...]) -> CausalEventRow:
    event, paths = row
    return CausalEventRow(
        id=event.id,
        title=event.title,
        occurred_on=event.occurred_on,
        first_seen_week=event.first_seen_week,
        paths=paths,
    )


def channel_of(row: tuple[Any, ...]) -> CausalChannelRow:
    channel, steps = row
    return CausalChannelRow(
        id=channel.id,
        name=channel.name,
        first_seen_week=channel.first_seen_week,
        steps=steps,
    )


class MarketCausalReadService:
    """인과 그래프를 읽어 응답 계약으로 준다."""

    def __init__(
        self,
        repository: MarketCausalReadRepository,
        graph_repository: CausalGraphReadRepository,
    ) -> None:
        self._repository = repository
        # **저장소가 둘이다.** 숫자·근거는 Postgres가 원본이고 모양은 Neo4j 투영이 낫다 —
        # 투영은 주를 넘어 노드를 공유해서 경로 목록이 못 잇는 사슬을 잇는다.
        self._graph = graph_repository

    async def paths(
        self,
        *,
        start: date,
        end: date,
        target_kinds: tuple[str, ...] = (),
        target_codes: tuple[str, ...] = (),
        event_id: int | None = None,
        limit: int = DEFAULT_LIMIT,
        offset: int = 0,
    ) -> CausalPathList:
        rows = await self._repository.path_rows(
            start=start,
            end=end,
            target_kinds=target_kinds,
            target_codes=target_codes,
            event_id=event_id,
            limit=limit,
            offset=offset,
        )
        return build_paths(rows, limit=limit, offset=offset)

    async def detail(self, path_id: int) -> CausalPathDetail:
        rows = await self._repository.detail_rows(path_id)
        if rows is None:
            raise UnknownPath(str(path_id))
        return build_detail(rows, path_id)

    async def week_graph(self, week: date) -> CausalGraph:
        """그 주의 투영. 그래프 DB가 없으면 `GraphOffline`이다."""
        if not self._graph.enabled:
            raise GraphOffline("neo4j is not configured for this process")
        return graph_of(self._graph.week_graph(week), week)

    async def target_chain(self, *, kind: str, code: str, limit: int = 50) -> CausalGraph:
        """대상 하나에서 **주를 넘어** 뻗은 사슬. 경로 목록으로는 못 만드는 답이다."""
        if not self._graph.enabled:
            raise GraphOffline("neo4j is not configured for this process")
        return graph_of(self._graph.target_chain(kind=kind, code=code, limit=limit), None)

    async def events(
        self, *, start: date, end: date, limit: int = DEFAULT_LIMIT, offset: int = 0
    ) -> CausalEventList:
        rows, has_more = await self._repository.event_rows(
            start=start, end=end, limit=limit, offset=offset
        )
        return CausalEventList(
            items=tuple(event_of(row) for row in rows),
            limit=limit,
            offset=offset,
            has_more=has_more,
        )

    async def channels(
        self, *, start: date, end: date, limit: int = DEFAULT_LIMIT, offset: int = 0
    ) -> CausalChannelList:
        rows, has_more = await self._repository.channel_rows(
            start=start, end=end, limit=limit, offset=offset
        )
        return CausalChannelList(
            items=tuple(channel_of(row) for row in rows),
            limit=limit,
            offset=offset,
            has_more=has_more,
        )


__all__ = [
    "GraphOffline",
    "MarketCausalReadService",
    "UnknownPath",
    "build_detail",
    "build_paths",
    "channel_of",
    "event_of",
    "evidence_of",
    "graph_of",
    "path_of",
]
